lib/launchers/ocm.rb

```diff
@@ -0,0 +1,190 @@
+        unless @cloud_opts
+          raise "Cannot find cloud '#{cloud}' defined in '#{service_name}'"
+        end
+      end
+    end
+
+    # @param service_name [String] the service name of this openstack instance
+    #   to lookup in configuration
+    def default_opts(service_name)
+      return  conf[:services, service_name.to_sym]
+    end
+
+    def to_seconds(string)
+      regex_m = /^(\d+)\s*(m|min|minutes|mins)+$/
+      regex_h = /^(\d+)\s*(h|hour|hours|hrs)+$/
+      regex_d = /^(\d+)\s*(d|day|days)+$/
+      regex_w = /^(\d+)\s*(w|week|weeks|wks)+$/
+      if string.match(regex_m)
+        return string.match(regex_m)[1].to_i * 60
+      elsif string.match(regex_h)
+        return string.match(regex_h)[1].to_i * 60 * 60
+      elsif string.match(regex_d)
+        return string.match(regex_d)[1].to_i * 24 * 60 * 60
+      elsif string.match(regex_w)
+        return string.match(regex_w)[1].to_i * 7 * 24 * 60 * 60
+      else
+        raise "Cannot convert '#{string}' to seconds!"
+      end
+    end
+
+    # create a json which specifies OSD cluster
+    # in the future we plan to move the logic into the script 'osd-provision.sh'
+    def generate_json(name)
+      json_data = {
+        "name" => name,
+        "managed" => true,
+        "multi_az" => false,
+        "byoc" => false
+      }
+
+      if @multi_az
+        json_data.merge!({"multi_az" => @multi_az})
+      end
+
+      if @region
+        json_data.merge!({"region" => {"id" => @region}})
+      end
+
+      if @version
+        json_data.merge!({"version" => {"id" => "openshift-v#{@version}"}})
+      end
+
+      if @nodes
+        json_data.merge!({"nodes" => {"compute" => @nodes.to_i}})
+      end
+
+      if @lifespan
+        expiration = Time.now + to_seconds(@lifespan)
+        json_data.merge!({"expiration_timestamp" => expiration.strftime("%Y-%m-%dT%H:%M:%SZ")})
+      end
+
+      if @cloud_opts
+        case @cloud_opts[:cloud_type]
+        when "aws"
+          aws = Amz_EC2.new(service_name: @cloud)
+          json_data.merge!({"aws" => {"access_key_id":aws.access_key, "secret_access_key":aws.secret_key, "account_id":aws.account_id}})
+          json_data.merge!({"byoc" => true})
+        end
+      end
+
+      return json_data.to_json
+    end
+
+    # download the script 'osd-provision.sh' which takes care of the OSD installation/uninstallation
+    def download_osd_script
+      osd_repo_uri = ENV['GIT_OSD_URI'] || 'https://gitlab.cee.redhat.com/mk-bin-packing/mk-performance-tests.git'
+      osd_repo_dir = File.join(Dir.tmpdir, 'osd_repo')
+      FileUtils.rm_rf(osd_repo_dir)
+      git = BushSlicer::Git.new(uri: osd_repo_uri, dir: osd_repo_dir)
+      git.clone
+      osd_script = File.join(osd_repo_dir, 'scripts', 'osd-provision.sh')
+      if !File.exists?(osd_script)
+        rais "Cannot find #{osd_script}"
+      end
+      return osd_script
+    end
+
+    def shell(cmd)
+      if !system(cmd)
```

Comment:
This might be fine but to allow catching signals properly in Jenkins Job abort, it might be needed to use `Host.localhost.exec_raw()` method. I'm not sure yet if it makes any difference.

Reply:
Yes, please use `Host.localhost.exec()` otherwise there are delays with with signal handling.
```
res = Host.localhost.exec(command, arg1, arg2, ..., env: env, single: true, stderr: :stdout, stdout: STDOUT, timeout: 3600)
if res[:success]
  ...
else
  ...
end
```

You can skip `env` if you don't want to change env. You can pass a single argument with full bash command I think as you are passing to `system` atm.

Reply:
This will run command in workdir.

Reply:
At the end I figured out that th esignaling flakiness I have been observing were caused by the fact that Jenkins sends SIGTERM to all processes instead of the parent process and rely on signal forwarding. That's why using `system()` is alright. But it will run in whatever is current dir while I prefer to run all processes in the localhost workdir for consistency.

Reply:
I use `Host.localhost.exec(` as you have suggested